### PR TITLE
[16.0][MIG] account_invoice_alternate_payer

### PR DESCRIPTION
**Pull request by ChrisOForgeFlow**
@ForgeFlow

Removed tests:

- Onchange methods on moves don't create all lines in same way than older versions, only when post is done system recompute values according display type of lines
- Payment state is a computed field on account.move can't be written

### CODE REVIEW

**Review by gerard-vacas — APPROVED** [review]
LGTM
Tested regular and refund invoices in both customer and vendor

**Comment by gerard-vacas** [comment]
Hi @ChrisOForgeFlow 

Could you force the runboat rebuild so the module can get another approval?

Thank you!

**Comment by ChrisOForgeFlow** [comment]
> Hi @ChrisOForgeFlow
> 
> Could you force the runboat rebuild so the module can get another approval?
> 
> Thank you!

Done!

**Comment by ghost** [comment]
Hi @ChrisOForgeFlow, please improve the module's coverage. Thank you!

File: account_invoice_alternate_payer/migrations/13.0.1.0.0/post-migration.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            account_account as aa,
+            account_account_type as aat
+        WHERE am.alternate_payer_id IS NOT NULL
+            AND am.id = aml.move_id
+            AND am.state = 'draft'
+            AND aat.type in ('receivable', 'payable')
+            AND aa.id = aml.account_id
+            AND aat.id = aa.user_type_id
+        """,
+    )
```
Comment (ghost) [inline account_invoice_alternate_payer/migrations/13.0.1.0.0/post-migration.py]: Hello, is this file still necessary?

File: account_invoice_alternate_payer/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+            existing_key_fname,
+            needed_vals_fname,
+            needed_dirty_fname,
+            line_type,
+            container,
+        ):
+            if line_type == "payment_term" and container.get("records").filtered(
+                lambda x: x.alternate_payer_id
+            ):
+                for invoice in container.get("records").filtered(
```
Comment (ghost) [inline account_invoice_alternate_payer/models/account_move.py]: Could you store `container.get("records").filtered(lambda x: x.alternate_payer_id)` in a variable so that you don't have to recalculate it on the `for` loop?

File: account_invoice_alternate_payer/models/account_move.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            if move.is_inbound():
+                domain.extend([("credit", ">", 0), ("debit", "=", 0)])
+                type_payment = _("Outstanding credits")
+            else:
+                domain.extend([("credit", "=", 0), ("debit", ">", 0)])
+                type_payment = _("Outstanding debits")
+            info = {"title": "", "outstanding": True, "content": [], "move_id": move.id}
+            lines = self.env["account.move.line"].search(domain)
+            currency_id = move.currency_id
+            if len(lines) != 0:
```
Comment (ghost) [inline account_invoice_alternate_payer/models/account_move.py]: Innecessary `if` . In case `lines` is empty, it will not enter the `for` loop.

File: account_invoice_alternate_payer/models/account_move.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+        # CHECK ME: this change to commercial partner when find difference between main partner
+        # https://github.com/odoo/odoo/blob/16.0/addons/account/models/account_move.py#L3431
+        if "partner_id" in values and len(values.keys()) == 1:
+            lines_to_skip = self.filtered(
+                lambda x: x.move_id.alternate_payer_id
+                and x.display_type == "payment_term"
+                and x.partner_id == x.move_id.alternate_payer_id
+            )
+            return super(AccountMoveLine, self - lines_to_skip).write(values)
+        return super(AccountMoveLine, self).write(values)
```
Comment (ghost) [inline account_invoice_alternate_payer/models/account_move.py]: Innecessary arguments on `super()`.

**Review by ghost — CHANGES_REQUESTED** [review]

File: account_invoice_alternate_payer/migrations/13.0.1.0.0/post-migration.py (hunk 1 above)
Comment (ChrisOForgeFlow) [reply account_invoice_alternate_payer/migrations/13.0.1.0.0/post-migration.py]: Removed

File: account_invoice_alternate_payer/models/account_move.py (hunk 2 above)
Comment (ChrisOForgeFlow) [reply account_invoice_alternate_payer/models/account_move.py]: Done!

File: account_invoice_alternate_payer/models/account_move.py (hunk 3 above)
Comment (ChrisOForgeFlow) [reply account_invoice_alternate_payer/models/account_move.py]: Change to loop over for only

File: account_invoice_alternate_payer/models/account_move.py (hunk 4 above)
Comment (ChrisOForgeFlow) [reply account_invoice_alternate_payer/models/account_move.py]: Done!

**Review by ghost — APPROVED** [review]
LGTM!

**Review by SoniaViciana — APPROVED** [review]
Functional review LGTM

**Comment by ghost** [comment]
Hi @pedrobaeza, 

Could you review this?

Thank you!

**Comment by pedrobaeza** [comment]
Merging blindly according reviews:

/ocabot migration account_invoice_alternate_payer
/ocabot merge nobump
